<a href="https://colab.research.google.com/github/Francisss3/Data-Engineering/blob/main/BlogPlatform_Schema_index_Lab_Data_engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
pip install pymongo

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 19.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 313.6/313.6 kB 18.1 MB/s eta 0:00:00


In [3]:
### Schema Design and Indexing in MongoDB
# Jupyter Notebook Skeleton

# --- Part 1: Schema Design ---
# Decisions:
# 1. Users in its own collection because frequent lookups by user metadata.
# 2. Posts in its own collection; embed small arrays for comments and tags initially.
#    - Comments: embed for simplicity (fast read of post+comments), but refactor later if comments grow large.
#    - Tags: embed as simple strings; full tag documents in separate `tags` collection for analytics.

# --- Part 2: Implement the Schema ---

# Cell 1: Imports & MongoDB Connection
from pymongo.mongo_client import MongoClient
from pymongo.server_api import ServerApi
import certifi

In [4]:
username = "charbelfrancis03"
password = "hUmsVOJeW3LRlLZ6"
uri = f"mongodb+srv://{username}:{password}@cluster0.8otd0br.mongodb.net/?retryWrites=true&w=majority"
client = MongoClient(uri, server_api=ServerApi('1'), tls=True, tlsCAFile=certifi.where())

db = client.blog_platform
users = db.users
posts = db.posts
comments = db.comments      # used later for refactor
tags = db.tags

In [5]:
# Insert Sample Users
db.users.delete_many({})
users.insert_many([
    {"_id": 1, "name": "Alice", "email": "alice@example.com"},
    {"_id": 2, "name": "Bob",   "email": "bob@example.com"}
])


InsertManyResult([1, 2], acknowledged=True)

In [6]:
# Insert Sample Posts (with embedded comments & tags)
db.posts.delete_many({})
posts.insert_many([
    {
        "_id": 101,
        "title": "How to Use MongoDB",
        "content": "This is a guide to using MongoDB.",
        "author_id": 1,
        "comments": [
            {"user_id": 2, "text": "Great post!", "timestamp": "2024-09-12T10:00:00Z"}
        ],
        "tags": ["MongoDB", "Database"]
    },
    {
        "_id": 102,
        "title": "Introduction to PyMongo",
        "content": "Connect Python to MongoDB using PyMongo.",
        "author_id": 2,
        "comments": [],
        "tags": ["Python", "MongoDB"]
    }
])


InsertManyResult([101, 102], acknowledged=True)

In [7]:
# Insert Tag Documents for Analytics
db.tags.delete_many({})
tags.insert_many([
    {"_id": "MongoDB", "description": "NoSQL database platform."},
    {"_id": "Database", "description": "General data storage topic."},
    {"_id": "Python", "description": "Programming language."}
])

InsertManyResult(['MongoDB', 'Database', 'Python'], acknowledged=True)

In [8]:
# --- Part 3: Indexing for Performance ---

# Fetch all posts by author and create an index on author_id
print("Before indexing, explain plan:")
print(posts.find({"author_id": 1}).explain())
# Create index
db.posts.create_index([("author_id", 1)])
print("After indexing, explain plan:")
print(posts.find({"author_id": 1}).explain())

Before indexing, explain plan:
{'explainVersion': '1', 'queryPlanner': {'namespace': 'blog_platform.posts', 'parsedQuery': {'author_id': {'$eq': 1}}, 'indexFilterSet': False, 'queryHash': 'E3A0FA33', 'planCacheShapeHash': 'E3A0FA33', 'planCacheKey': 'BC11A40F', 'optimizationTimeMillis': 0, 'maxIndexedOrSolutionsReached': False, 'maxIndexedAndSolutionsReached': False, 'maxScansToExplodeReached': False, 'prunedSimilarIndexes': False, 'winningPlan': {'isCached': False, 'stage': 'COLLSCAN', 'filter': {'author_id': {'$eq': 1}}, 'direction': 'forward'}, 'rejectedPlans': []}, 'executionStats': {'executionSuccess': True, 'nReturned': 1, 'executionTimeMillis': 0, 'totalKeysExamined': 0, 'totalDocsExamined': 2, 'executionStages': {'isCached': False, 'stage': 'COLLSCAN', 'filter': {'author_id': {'$eq': 1}}, 'nReturned': 1, 'executionTimeMillisEstimate': 0, 'works': 3, 'advanced': 1, 'needTime': 1, 'needYield': 0, 'saveState': 0, 'restoreState': 0, 'isEOF': 1, 'direction': 'forward', 'docsExamined

In [9]:
# Query comments by user and index comments.user_id
# (On embedded comments, use a covered index on posts.comments.user_id via a multikey index.)
db.posts.create_index([("comments.user_id", 1)])
print("Comments by user 2:")
results = posts.find({"comments.user_id": 2}, {"_id":0, "title":1, "comments.$":1})
for doc in results:
    print(doc)

Comments by user 2:
{'title': 'How to Use MongoDB', 'comments': [{'user_id': 2, 'text': 'Great post!', 'timestamp': '2024-09-12T10:00:00Z'}]}


In [10]:
# --- Part 4: Refactoring for Performance ---
# When comments grow large, embedding will bloat posts doc size and slow updates.
# Instead, store comments in a separate `comments` collection and reference posts.

# Move embedded comments into separate collection
existing = list(posts.find({"comments": {"$exists": True}}, {"_id":1, "comments":1}))

# Clear and recreate comments collection
comments.delete_many({})
for post in existing:
    for c in post.get("comments", []):
        comment_doc = {
            "post_id": post["_id"],
            "user_id": c["user_id"],
            "text": c["text"],
            "timestamp": c["timestamp"]
        }
        comments.insert_one(comment_doc)
# Remove embedded comments from posts
db.posts.update_many({}, {"$unset": {"comments": ""}})
print("Refactored comments into separate collection. Total comments:", comments.count_documents({}))

# Create index on comments.user_id and comments.post_id
comments.create_index([("user_id", 1)])
comments.create_index([("post_id", 1)])

Refactored comments into separate collection. Total comments: 1


'post_id_1'

In [11]:
# Test refactored queries
print("Comments by user 2 after refactor:")
for doc in comments.find({"user_id": 2}, {"_id":0, "post_id":1, "text":1}):
    print(doc)

print("Posts by author 1 still works:")
for p in posts.find({"author_id": 1}, {"_id":0, "title":1}):
    print(p)


Comments by user 2 after refactor:
{'post_id': 101, 'text': 'Great post!'}
Posts by author 1 still works:
{'title': 'How to Use MongoDB'}


In [15]:
# Part 5a: Benchmark the “unindexed” case
posts.drop_indexes()           # remove every index except the default _id
%timeit posts.count_documents({"author_id": 1})


210 ms ± 217 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [16]:
# Part 5b: Benchmark the “indexed” case
posts.create_index([("author_id", 1)])
%timeit posts.count_documents({"author_id": 1})


210 ms ± 210 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)


### Performance Summary

- **Without index** (`posts.drop_indexes()`):  
  ~210 ms per loop  
- **With index** (`posts.create_index([("author_id",1)])`):  
  ~210 ms per loop  

**Analysis:**  
Because our collection only has three small documents, both scans take about the same time—and the index isn’t used meaningfully. In a real‐world blog with thousands or millions of posts, the indexed query would avoid a full collection scan and be orders of magnitude faster, while the unindexed version would slow down proportionally with data size.

> *Tip:* Run `posts.find({"author_id":1}).explain()` to confirm whether MongoDB is using the index in its query plan.
